### PR TITLE
[SMTChecker] Support for mapping

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #5387 

This PR adds support to R-d mapping access and 1-d mapping assignment.

### CODE REVIEW

**Comment by axic** [comment]
On Mac this is crashing:
```
Testing soljson via the fuzzer...
scripts/../test/cmdlineTests.sh: line 312: 28380 Segmentation fault: 11  "$REPO_ROOT"/build/test/tools/solfuzzer --quiet < "$f"
Fuzzer failed on:
pragma experimental SMTChecker;

contract C
{
	mapping (uint => mapping (uint => mapping (uint => uint))) map;
	function f(uint x) public {
		x = 42;
		map[13][14][15] = 42;
		assert(x == map[13][14][15]);
	}
}
Commandline tests FAILED
Exited with code 1
```

File: libsolidity/formal/SymbolicTypes.cpp [hunk 1]
```diff
@@ -45,7 +45,12 @@ smt::SortPointer dev::solidity::smtSort(Type const& _type)
 	}
 	case smt::Kind::Array:
 	{
-		solUnimplementedAssert(false, "Invalid type");
+		if (isMapping(_type.category()))
+		{
+			auto mapType = dynamic_cast<MappingType const*>(&_type);
+			solAssert(mapType, "");
+			return make_shared<smt::ArraySort>(smtSort(*mapType->keyType()), smtSort(*mapType->valueType()));
+		}
```
Comment (axic) [inline libsolidity/formal/SymbolicTypes.cpp:60]: I prefer to have an assertion here as opposed to a fallthrough.

**Comment by axic** [comment]
```
  * frame #0: 0x00000001004ec1e7 solc`dev::solidity::smt::Sort::operator==(dev::solidity::smt::Sort const&) const + 23
    frame #1: 0x00000001004c5b3a solc`dev::solidity::smt::Expression::store(dev::solidity::smt::Expression, dev::solidity::smt::Expression, dev::solidity::smt::Expression) + 2122
    frame #2: 0x00000001004c46a8 solc`dev::solidity::SMTChecker::arrayAssignment(dev::solidity::Assignment const&) + 2136
    frame #3: 0x00000001004c3b07 solc`dev::solidity::SMTChecker::endVisit(dev::solidity::Assignment const&) + 1959
    frame #4: 0x0000000100225cc2 solc`dev::solidity::Assignment::accept(dev::solidity::ASTConstVisitor&) const + 146
    frame #5: 0x00000001002253ea solc`dev::solidity::ExpressionStatement::accept(dev::solidity::ASTConstVisitor&) const + 122

```

**Comment by leonardoalt** [comment]
Do you think it could be related to `operator==` not being virtual?

**Comment by axic** [comment]
Added assertions. It seems that `arrayAssignment` pushes an invalid expression for element.

**Comment by leonardoalt** [comment]
Crash fixed. Ready for review again.

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -248,13 +247,72 @@ void SMTChecker::endVisit(Assignment const& _assignment)
...
 	else
 		m_errorReporter.warning(
 			_assignment.location(),
 			"Assertion checker does not yet implement such assignments."
 		);
 }
 
+void SMTChecker::arrayAssignment(Assignment const& _assignment)
+{
+	auto left = dynamic_cast<IndexAccess const*>(&_assignment.leftHandSide());
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: Why not do a dynamic cast by reference?

File: libsolidity/formal/SMTChecker.cpp [hunk 3]
```diff
@@ -248,13 +247,72 @@ void SMTChecker::endVisit(Assignment const& _assignment)
...
+	solAssert(left, "");
+
+	vector<smt::Expression> stores;
+	stores.emplace_back(expr(_assignment.rightHandSide()));
+	bool error = false;
+	// This loop handles multi-dimensional arrays.
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: What about things like `a[2].x[4][3]`?

File: libsolidity/formal/SMTChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: `left` is reassigned later when used as a loop iterator

File: libsolidity/formal/SMTChecker.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: What's the type of `a`?
I should perhaps phrase it differently. Solidity arrays are not supported yet, I meant SMT arrays when writing that, to support Solidity mappings.

Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp]: OK, I see - but does it work or not?

I'm asking more in light of a planned architecture - you cannot just stay with mappings of mappings in the long run, you have to introduce some abstract lvalue.

Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: It would work by adding MemberAccess there.
But I see your point. I can rework it next week.

**Comment by leonardoalt** [comment]
@chriseth Updated

**Comment by chriseth** [comment]
Please add some tests that check for unsupported key or value types in mappings.

File: libsolidity/formal/SMTChecker.h [hunk 4]
```diff
@@ -92,6 +94,8 @@ class SMTChecker: private ASTConstVisitor
 	/// if available and inlines the return value.
 	void inlineFunctionCall(FunctionCall const&);
 
+	void encodeArrayAssignment(IndexAccess const&, std::shared_ptr<SymbolicVariable> const&);
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.h]: No parameter names?

File: libsolidity/formal/SMTChecker.cpp [hunk 5]
```diff
@@ -251,6 +250,15 @@ void SMTChecker::endVisit(VariableDeclarationStatement const& _varDecl)
 		);
 }
 
+bool SMTChecker::visit(Assignment const& _assignment)
+{
+	m_assignmentLeftHand = true;
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: What about `m[x[3] = 2] = 8`? :)

File: libsolidity/formal/SMTChecker.cpp [hunk 6]
```diff
@@ -626,6 +643,100 @@ bool SMTChecker::visit(MemberAccess const& _memberAccess)
...
+			// multi-dimensional access.
+			// This way we only ask a model for the outermost access.
+			if (m_uninterpretedTerms.count(innerAccess))
+				m_uninterpretedTerms.erase(innerAccess);
+		}
+		else
+			error = true;
+	}
+	else
+		error = true;
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: Might also initialize `error = true;`

**Review by chriseth — COMMENTED** [review]
Apart from the comments (and the additional tests we discussed offline) this looks good, but I have the impression that a manual implementation of `visit(Assignment)`, that has a "will be assigned" parameter and returns the expression to assign to might be easier to implement than sticking with the fixed visitor pattern.

**Comment by chriseth** [comment]
Needs rebase.

**Comment by leonardoalt** [comment]
@chriseth updated.
Added a test with an unsupported mapping key type. It doesn't crash/throw, which is good, but it does say "expression undefined for smt solver". I already have an issue for cleaning up these cases later.

**Comment by chriseth** [comment]
Please add a test for
`assert(x == y); assert(map[x] == map[y]);`

**Review by chriseth — CHANGES_REQUESTED** [review]
Fine apart from the requested test.

**Comment by chriseth** [comment]
Ah another idea: Does passing mappings via internal function calls work? Do assignments for variables of mapping type work?

```
contract c {
  mapping(uint => uint) x;
  mapping(uint => uint) y;
  function f(bool cond, uint s) public {
    mapping(uint => uint) a = cond ? x : y;
    x[2] = 1;
    y[2] = 2;
    a[2] = 3;
    if (cond)
      assert(a[2] == x[2] && a[2] != y[2]);
    else
      assert(a[2] == y[2] && a[2] != x[2]);
  }
}
```

**Comment by leonardoalt** [comment]
Nice ideas. Added tests for those. The examples from the last comment yield false positives since the SMTChecker does not consider aliasing ;)

**Comment by chriseth** [comment]
@leonardoalt so can "not supporting aliasing" cause any harm? If not, why not?

File: test/libsolidity/smtCheckerTests/types/mapping_as_parameter_1.sol [hunk 7]
```diff
@@ -0,0 +1,14 @@
+pragma experimental SMTChecker;
+
+contract c {
+	mapping(uint => uint) x;
+	function f(mapping(uint => uint) storage map, uint index, uint value) internal {
+		map[index] = value;
+	}
+	function g(uint a, uint b) public {
+		f(x, a, b);
+		assert(x[a] == b);
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/types/mapping_as_parameter_1.sol:11]: can you add a comment that this is not yet supported?

File: test/libsolidity/smtCheckerTests/types/mapping_as_local_var_1.sol [hunk 8]
```diff
@@ -0,0 +1,20 @@
+pragma experimental SMTChecker;
+
+contract c {
+	mapping(uint => uint) x;
+	mapping(uint => uint) y;
+	function f(bool cond) public {
+		mapping(uint => uint) storage a = cond ? x : y;
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/types/mapping_as_local_var_1.sol:7]: Please also add a comment here.

**Comment by leonardoalt** [comment]
Thinking about it, it is harmful (as in false negatives):
```
    mapping (uint => uint) a;
    mapping (uint => uint) b;
    function f() public {
        require(a[1] == b[1]);
        mapping (uint => uint) storage c = a;
        c[1] = 2;
        assert(a[1] == b[1]);
    }
```

**Comment by leonardoalt** [comment]
My suggestion: we should try to support it and whenever it's not possible we clear all mapping knowledge.

**Comment by chriseth** [comment]
Sounds good!

**Comment by chriseth** [comment]
What does this mean for the scope of this PR?

**Comment by leonardoalt** [comment]
I'll erase all mapping knowledge if there's an assignment or parameter of mapping type in this PR.

**Review by chriseth — APPROVED** [review]

**Pull request merged**